### DISCUSSION TITLE
Typer API Documentation (or: Where do I find the arguments to Typer())?

### DISCUSSION THREAD

**Discussion by linsomniac** [fastapi/typer#701] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the Typer documentation, with the integrated search.
- [X] I already searched in Google "How to X in Typer" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to Typer but to [Click](https://github.com/pallets/click).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
N/A, this is a documentation discussion.
```

### Description

I really like Typer, but one thing I struggle with is the "narrative style" tutorial documentation; it's really good for getting started, but I really struggle with using it to answer questions like "What documentation options do I have in the "Typer()" call?"  I end up clicking all around and hunting for something that answers my question.  For example, searching for "epilog" in the typer documentation doesn't list it as being an option to Typer(), but it is.

What I'm thinking is:

- Adding Docstrings to Typer() and friends, including documenting arguments.
- Adding an "API Docs" or similar to the mkdocs nav.
- Adding "docs/api/index.md" (or possibly others) that uses mkdocstrings to pull in the documentation from the docstrings.

Thoughts?  Am I missing some documentation?

### Operating System

Linux

### Typer Version

0.9.0

### Python Version

3.11

**Comment by 9999years** [comment]
I would also like this. I can't find anywhere that lists or explains the `typer.Option` arguments, for instance, and many of them are not documented in the tutorial-style documentation either. (What's `allow_from_autoenv`? `expose_value`? `shell_complete` vs. `autocompletion`?)

**Reply by svlandeg** [reply]
FYI - on the topic of `shell_complete` vs `autocompletion` - we've decided to only keep the latter and not support the former, to avoid confusion. Cf #949 and #974.

**Comment by svlandeg** [comment]
I see your point - I'll label this as a feature request.

In the meantime, while it's not fully complete yet, you might enjoy looking at this user contribution: https://github.com/tiangolo/typer/discussions/634 -> [Cheat sheet](https://gist.github.com/harkabeeparolus/a6e18b1f4f4f938f450090c5e7523f68)

**Comment by gcr** [comment]
Strong agree.These docs and others (looking at you, sqlmodel :) ) seem to lack "readthedocs"-style API references for each class, which is a significant problem in my book.

It's great that there's high-quality material for beginners -- I wish more libraries had this -- but it also makes complex issues with this library impossible to debug.

(Thx for a great library btw!)

**Comment by mihaic** [comment]
Related issue:
https://github.com/fastapi/typer/issues/221

**Comment by gcr** [comment]
if one of us outsiders were to just start to write some docstrings and send a pull request, would that be considered? Or would the contributor team prefer to coordinate this effort when the time is right?

**Comment by svlandeg** [comment]
Hi all! I just discussed this with Tiangolo and we agreed I'll work on this. I'll prioritize this in the next few weeks. If there's an easy way to help out (while still ensuring things are coordinated), I'll let you know here. Thanks!

**Comment by svlandeg** [comment]
Draft PR: https://github.com/fastapi/typer/pull/1504

**Comment by svlandeg** [comment, marked as answer]
The typer reference docs are now available here: https://typer.tiangolo.com/reference/

**Reply by gcr** [reply]
oh these look awesome!! Thank you!
